### PR TITLE
[PSRule] Fix Keyvault issues 

### LINKED ISSUE #3987: [PSRule] Fix Keyvault issues
PsRule validation fails to expand KV bicep modules, due to an issue in private endpoint:
failed to expand bicep source '/home/runner/work/ResourceModules/ResourceModules/modules/key-vault/vault/.test/common/main.test.bicep'. Exception calling "GetBicepResources" with "2" argument(s): "Unable to expand resources because the source file '/home/runner/work/ResourceModules/ResourceModules/modules/key-vault/vault/.test/common/main.test.bicep' was not valid. An error occurred evaluating expression '[format('{0}-PrivateEndpoint-PrivateDnsZoneGroup', uniqueString(deployment().name, parameters('location')))]' line 1951. The argument 'additional parameters as needed' for 'UniqueString' is not a valid string."Show less

This PR solve the above issue and the following requirements for KV deployment:
Azure.KeyVault.Firewall | ***kvvmin002 | Key Vault should only accept explicitly allowed traffic.
-- | -- | --
Azure.KeyVault.RBAC | ***kvvmin002 | Key Vaults should use Azure RBAC as the authorization system for the data plane.

### PR DESCRIPTION
**Pull request by elanzel**
# Description
Fix of:
Exception calling "GetBicepResources" with "2" argument(s): "Unable to expand resources because the source file '/home/runner/work/ResourceModules/ResourceModules/modules/key-vault/vault/.test/common/main.test.bicep' was not valid.
Azure.KeyVault.Firewall | ***kvvmin002 | Key Vault should only accept explicitly allowed traffic.
Azure.KeyVault.RBAC | ***kvvmin002 | Key Vaults should use Azure RBAC as the authorization system for the data plane.

closes #3987 

### Pipeline references
| Pipeline |
| - |
[KeyVault - Vaults](https://github.com/Azure/ResourceModules/actions/workflows/ms.keyvault.vaults.yml) |

## Type of Change

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: modules/key-vault/vault/.test/common/main.test.bicep [hunk 1]
```diff
@@ -64,42 +64,14 @@ module testDeployment '../../main.bicep' = {
   params: {
     enableDefaultTelemetry: enableDefaultTelemetry
     name: '${namePrefix}${serviceShort}002'
-    accessPolicies: [
```
Comment (AlexanderSehr) [inline modules/key-vault/vault/.test/common/main.test.bicep:67]: I'd veto against this change (we have to discuss this) as you're right that access policies are not PSRule/WAF aligned - but - without them here, we don't test the child resource.

My suggestion: Let's create a copy of the min test and call it 'access policies', and the access policies in there (and exclude the file from the PSRule test), and perform the change in the common test as per your suggestion.

Comment (elbatane) [reply modules/key-vault/vault/.test/common/main.test.bicep:67]: Already implemented

File: modules/key-vault/vault/.test/accesspolicies/main.test.bicep [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  location: location
+}
+
+module nestedDependencies 'dependencies.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-nestedDependencies'
+  params: {
+    managedIdentityName: 'dep-${namePrefix}-msi-${serviceShort}'
+    virtualNetworkName: 'dep-${namePrefix}-vnet-${serviceShort}'
+
```
Comment (AlexanderSehr) [inline modules/key-vault/vault/.test/accesspolicies/main.test.bicep:40]:
```suggestion
```

File: modules/key-vault/vault/.test/min/main.test.bicep [hunk 3]
```diff
@@ -43,5 +43,6 @@ module testDeployment '../../main.bicep' = {
     name: '${namePrefix}${serviceShort}002'
     // Only for testing purposes
     enablePurgeProtection: false
+    enableRbacAuthorization: true
```
Comment (eriqua) [inline modules/key-vault/vault/.test/min/main.test.bicep:46]: min tests should only contain required parameters. I'd suggest to remove this and instead change the `enableRbacAuthorization` default value to `true` in the main.bicep template

```suggestion
```

File: modules/key-vault/vault/.test/common/main.test.bicep [hunk 4]
```diff
@@ -64,42 +64,14 @@ module testDeployment '../../main.bicep' = {
...
-    ]
+
     diagnosticStorageAccountId: diagnosticDependencies.outputs.storageAccountResourceId
     diagnosticWorkspaceId: diagnosticDependencies.outputs.logAnalyticsWorkspaceResourceId
     diagnosticEventHubAuthorizationRuleId: diagnosticDependencies.outputs.eventHubAuthorizationRuleId
     diagnosticEventHubName: diagnosticDependencies.outputs.eventHubNamespaceEventHubName
     // Only for testing purposes
     enablePurgeProtection: false
-    enableRbacAuthorization: false
+    enableRbacAuthorization: true
```
Comment (eriqua) [inline modules/key-vault/vault/.test/common/main.test.bicep:74]: As above, I'd suggest to remove this line and instead change the enableRbacAuthorization default value to true in the main.bicep template

File: modules/key-vault/vault/.test/accesspolicies/main.test.bicep [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}'
+  params: {
+    enableDefaultTelemetry: enableDefaultTelemetry
+    name: '${namePrefix}${serviceShort}002'
+    diagnosticStorageAccountId: diagnosticDependencies.outputs.storageAccountResourceId
+    diagnosticWorkspaceId: diagnosticDependencies.outputs.logAnalyticsWorkspaceResourceId
+    diagnosticEventHubAuthorizationRuleId: diagnosticDependencies.outputs.eventHubAuthorizationRuleId
+    diagnosticEventHubName: diagnosticDependencies.outputs.eventHubNamespaceEventHubName
+    // Only for testing purposes
```
Comment (eriqua) [inline modules/key-vault/vault/.test/accesspolicies/main.test.bicep:71]: Should we remove this comment?

File: modules/key-vault/vault/README.md [hunk 6]
```diff
@@ -755,7 +868,7 @@ module vault './key-vault/vault/main.bicep' = {
 </details>
 <p>
 
-<h3>Example 2: Min</h3>
+<h3>Example 3: Min</h3>
```
Comment (eriqua) [inline modules/key-vault/vault/README.md:871]: Please regenerate readmes after applying suggestions

File: modules/key-vault/vault/.test/min/main.test.bicep (hunk 3 above)
Comment (elanzel) [reply modules/key-vault/vault/.test/min/main.test.bicep]: All suggestion committed, readme and main.json updated

File: modules/key-vault/vault/.test/accesspolicies/main.test.bicep (hunk 5 above)
Comment (elanzel) [reply modules/key-vault/vault/.test/accesspolicies/main.test.bicep]: removed

File: modules/key-vault/vault/README.md [hunk 7]
```diff
@@ -47,7 +47,7 @@ This module deploys a Key Vault.
...
+| `enableRbacAuthorization` | bool | `True` |  | Property that controls how data actions are authorized. When true, the key vault will use Role Based Access Control (RBAC) for authorization of data actions, and the access policies specified in vault properties will be ignored (warning: this is a preview feature). When false, the key vault will use the access policies specified in vault properties, and any policy stored on Azure Resource Manager will be ignored. If null or not specified, the vault is created with the default value of false. Note that management actions are always authorized with RBAC. |
```
Comment (eriqua) [inline modules/key-vault/vault/README.md:50]:
```suggestion
| `enableRbacAuthorization` | bool | `True` |  |Property that controls how data actions are authorized. When true, the key vault will use Role Based Access Control (RBAC) for authorization of data actions, and the access policies specified in vault properties will be ignored. When false, the key vault will use the access policies specified in vault properties, and any policy stored on Azure Resource Manager will be ignored. If null or not specified, the vault is created with the default value of false. Note that management actions are always authorized with RBAC.  |
```

File: modules/key-vault/vault/.test/accesspolicies/main.test.bicep [hunk 8]
```diff
@@ -12,7 +12,7 @@ param resourceGroupName string = 'ms.keyvault.vaults-${serviceShort}-rg'
 param location string = deployment().location
 
 @description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
-param serviceShort string = 'kvvaccesspolicy'
+param serviceShort string = 'kvvrbac'
```
Comment (eriqua) [inline modules/key-vault/vault/.test/accesspolicies/main.test.bicep:15]:
```suggestion
param serviceShort string = 'kvvac'
```

File: modules/key-vault/vault/main.bicep [hunk 9]
```diff
@@ -38,7 +38,7 @@ param enableSoftDelete bool = true
 param softDeleteRetentionInDays int = 90
 
 @description('Optional. Property that controls how data actions are authorized. When true, the key vault will use Role Based Access Control (RBAC) for authorization of data actions, and the access policies specified in vault properties will be ignored (warning: this is a preview feature). When false, the key vault will use the access policies specified in vault properties, and any policy stored on Azure Resource Manager will be ignored. If null or not specified, the vault is created with the default value of false. Note that management actions are always authorized with RBAC.')
```
Comment (eriqua) [inline modules/key-vault/vault/main.bicep:40]:
```suggestion
@description('Optional. Property that controls how data actions are authorized. When true, the key vault will use Role Based Access Control (RBAC) for authorization of data actions, and the access policies specified in vault properties will be ignored. When false, the key vault will use the access policies specified in vault properties, and any policy stored on Azure Resource Manager will be ignored. Note that management actions are always authorized with RBAC.')
```

File: modules/key-vault/vault/main.json [hunk 10]
```diff
@@ -84,7 +84,7 @@
     },
     "enableRbacAuthorization": {
       "type": "bool",
-      "defaultValue": false,
+      "defaultValue": true,
```
Comment (eriqua) [inline modules/key-vault/vault/main.json:87]: Please regen after applying suggestions

File: modules/key-vault/vault/README.md [hunk 11]
```diff
@@ -47,7 +47,7 @@ This module deploys a Key Vault.
 | `diagnosticWorkspaceId` | string | `''` |  | Resource ID of the diagnostic log analytics workspace. For security reasons, it is recommended to set diagnostic settings to send data to either storage account, log analytics workspace or event hub. |
 | `enableDefaultTelemetry` | bool | `True` |  | Enable telemetry via a Globally Unique Identifier (GUID). |
 | `enablePurgeProtection` | bool | `True` |  | Provide 'true' to enable Key Vault's purge protection feature. |
-| `enableRbacAuthorization` | bool | `False` |  | Property that controls how data actions are authorized. When true, the key vault will use Role Based Access Control (RBAC) for authorization of data actions, and the access policies specified in vault properties will be ignored (warning: this is a preview feature). When false, the key vault will use the access policies specified in vault properties, and any policy stored on Azure Resource Manager will be ignored. If null or not specified, the vault is created with the default value of false. Note that management actions are always authorized with RBAC. |
```
Comment (eriqua) [inline modules/key-vault/vault/README.md:50]: Please regen after applying suggestions

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**